engine/src/main/java/org/terasology/world/chunks/remoteChunkProvider/RemoteChunkProvider.java

```diff
@@ -24,105 +11,100 @@
 import org.terasology.world.chunks.event.BeforeChunkUnload;
 import org.terasology.world.chunks.event.OnChunkLoaded;
-import org.terasology.world.chunks.internal.GeneratingChunkProvider;
-import org.terasology.world.chunks.pipeline.AbstractChunkTask;
-import org.terasology.world.chunks.pipeline.ChunkGenerationPipeline;
+import org.terasology.world.chunks.pipeline.ChunkProcessingPipeline;
 import org.terasology.world.chunks.pipeline.ChunkTask;
+import org.terasology.world.chunks.pipeline.LightMergerChunkTaskProvider;
+import org.terasology.world.chunks.pipeline.tasks.DeflateChunkTask;
+import org.terasology.world.chunks.pipeline.tasks.GenerateInternalLightningChunkTask;
+import org.terasology.world.chunks.pipeline.tasks.NotifyChunkTask;
 import org.terasology.world.internal.ChunkViewCore;
 import org.terasology.world.internal.ChunkViewCoreImpl;
-import org.terasology.world.propagation.light.InternalLightProcessor;
-import org.terasology.world.propagation.light.LightMerger;
 
 import java.math.RoundingMode;
 import java.util.Collection;
-import java.util.Collections;
 import java.util.Comparator;
 import java.util.List;
 import java.util.Map;
 import java.util.concurrent.BlockingQueue;
 
 /**
- *
+ * Provides chunks received from remote source.
+ * <p>
+ * Loading/Unload chunks dependent on {@link org.terasology.network.Server}
+ * <p/>
+ * Produce events:
+ * <p>
+ * {@link OnChunkLoaded} when chunk received from server and processed.
+ * <p>
+ * {@link BeforeChunkUnload} when {@link org.terasology.network.Server} send invalidate chunk and chunk removing
  */
-public class RemoteChunkProvider implements ChunkProvider, GeneratingChunkProvider {
+public class RemoteChunkProvider implements ChunkProvider {
 
     private static final Logger logger = LoggerFactory.getLogger(RemoteChunkProvider.class);
-    private Map<Vector3i, Chunk> chunkCache = Maps.newHashMap();
-    private final BlockingQueue<Chunk> readyChunks = Queues.newLinkedBlockingQueue();
     private final BlockingQueue<Vector3i> invalidateChunks = Queues.newLinkedBlockingQueue();
-    private List<Chunk> sortedReadyChunks = Lists.newArrayList();
+    private Map<Vector3i, Chunk> chunkCache = Maps.newHashMap();
     private ChunkReadyListener listener;
     private EntityRef worldEntity = EntityRef.NULL;
 
     private BlockManager blockManager;
 
-    private ChunkGenerationPipeline pipeline;
-
-    private LightMerger<Chunk> lightMerger = new LightMerger<>(this);
+    private ChunkProcessingPipeline loadingPipeline;
 
     private LocalPlayer localPlayer;
 
     public RemoteChunkProvider(BlockManager blockManager, LocalPlayer localPlayer) {
         this.blockManager = blockManager;
         this.localPlayer = localPlayer;
-        pipeline = new ChunkGenerationPipeline(new ChunkTaskRelevanceComparator());
+        loadingPipeline = new ChunkProcessingPipeline(new ChunkTaskRelevanceComparator());
+
+        loadingPipeline.addStage(GenerateInternalLightningChunkTask::new)
```

Comment:
ahh, this is probably the important part of this PR, isn't it.

Okay, I can appreciate having some pipeline stages that have a consistent interface.

I may have some questions but it looks like I need to read ChunkProcessingPipeline.